dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/utils/JweCreator.java

```diff
@@ -0,0 +1,138 @@
+import com.ibm.ws.security.jwt.internal.BuilderImpl;
+import com.ibm.ws.security.jwt.internal.JwtTokenException;
+
+public class JweCreator {
+
+    private static final TraceComponent tc = Tr.register(JweCreator.class);
+
+    @FFDCIgnore({ Exception.class })
+    public static String createJweString(String jws, JwtData jwtData) throws Exception {
+        JweCreator signer = new JweCreator();
+        JwtConfig jwtConfig = jwtData.getConfig();
+        try {
+            JsonWebEncryption jwe = new JsonWebEncryption();
+            BuilderImpl builder = jwtData.getBuilder();
+            signer.setJweKeyData(jwe, builder, jwtConfig);
+            signer.setJweHeaders(jwe, builder, jwtConfig);
+            jwe.setPayload(jws);
+            return signer.getJwtString(jwe);
+        } catch (Exception e) {
+            String errorMsg = Tr.formatMessage(tc, "ERROR_BUILDING_SIGNED_JWE", new Object[] { jwtConfig.getId(), e });
+            throw new Exception(errorMsg, e);
+        }
+    }
+
+    void setJweKeyData(JsonWebEncryption jwe, BuilderImpl builder, JwtConfig jwtConfig) throws KeyStoreException, CertificateException, InvalidTokenException {
+        Key keyManagementKey = getKeyManagementKey(builder, jwtConfig);
+        if (keyManagementKey == null) {
+            String errorMsg = Tr.formatMessage(tc, "KEY_MANAGEMENT_KEY_NOT_FOUND", new Object[] { jwtConfig.getId(), jwtConfig.getKeyManagementKeyAlias(), jwtConfig.getTrustStoreRef() });
+            throw new KeyStoreException(errorMsg);
+        }
+        jwe.setKey(keyManagementKey);
+        setJweKidHeader(jwe, keyManagementKey);
+    }
+
+    Key getKeyManagementKey(BuilderImpl builder, JwtConfig jwtConfig) throws KeyStoreException, CertificateException, InvalidTokenException {
+        Key keyManagementKey = builder.getKeyManagementKey();
+        if (keyManagementKey == null) {
+            keyManagementKey = getKeyManagementKeyFromTrustStore(jwtConfig);
+        }
+        return keyManagementKey;
+    }
+
+    Key getKeyManagementKeyFromTrustStore(JwtConfig jwtConfig) throws KeyStoreException, CertificateException, InvalidTokenException {
+        String keyAlias = jwtConfig.getKeyManagementKeyAlias();
+        String trustStoreRef = jwtConfig.getTrustStoreRef();
+        return JwtUtils.getPublicKey(keyAlias, trustStoreRef);
+    }
+
+    void setJweKidHeader(JsonWebEncryption jwe, Key keyManagementKey) {
+        JwkKidBuilder kidbuilder = new JwkKidBuilder();
+        String keyId = kidbuilder.buildKeyId(keyManagementKey);
+        if (keyId != null) {
+            jwe.setKeyIdHeaderValue(keyId);
+        }
+    }
+
+    void setJweHeaders(JsonWebEncryption jwe, BuilderImpl builder, JwtConfig jwtConfig) {
+        jwe.setAlgorithmHeaderValue(getKeyManagementKeyAlgorithm(builder, jwtConfig));
+        jwe.setEncryptionMethodHeaderParameter(getContentEncryptionAlgorithm(builder, jwtConfig));
+        jwe.setHeader("typ", "JOSE");
```

Comment:
For reference only: This is to comply with section 4.1.11 https://tools.ietf.org/html/rfc7516#section-4.1.11 of the JWE specification.